### ISSUE TITLE
Custom coco format dataset

### ISSUE THREAD

**Issue by Egorundel** [pytorch/vision#8087]
Hello! Can you suggest the structure of this dataset? 
I want to use a custom dataset in coco format. 
But I need to know what folder/file structure is needed for training.

```python
def get_args_parser(add_help=True):
    import argparse

    parser = argparse.ArgumentParser(description="PyTorch Detection Training", add_help=add_help)

    parser.add_argument("--data-path", default="/datasets01/COCO/022719/", type=str, help="dataset path")
    parser.add_argument(
        "--dataset",
        default="coco",
        type=str,
        help="dataset name. Use coco for object detection and instance segmentation and coco_kp for Keypoint detection",
    )
```

**Comment by pmeier** [comment]
The relevant code is here

https://github.com/pytorch/vision/blob/f69eee6108cd047ac8b62a2992244e9ab3c105e1/references/detection/train.py#L43-L49

https://github.com/pytorch/vision/blob/f69eee6108cd047ac8b62a2992244e9ab3c105e1/references/detection/coco_utils.py#L200-L210

Meaning, under the `--data-path` directory, you need two have three folders:

- `train2017`
- `val2017`
- `annotations`

Both `train2017` and `val2017` contain the images. The annotations folder contains two files, namely `instances_train2017.json` and `instances_val2017.json`. The format of the annotation files can be found [here (Object Detection)](https://cocodataset.org/#format-data)

**Comment by Egorundel** [comment]
@pmeier Hm... My json's have everything you need to use... 

Here is an example of my `instances_val.json`:
```
{
  "categories": [
    {
      "id": 0,
      "name": "man1",
      "supercategory": ""
    },
    {
      "id": 1,
      "name": "man2",
      "supercategory": ""
    },
    {
      "id": 2,
      "name": "man3",
      "supercategory": ""
    },
    {
      "id": 3,
      "name": "man4",
      "supercategory": ""
    },
    {
      "id": 4,
      "name": "man5",
      "supercategory": ""
    },
    {
      "id": 5,
      "name": "man6",
      "supercategory": ""
    },
    {
      "id": 6,
      "name": "man7",
      "supercategory": ""
    },
    {
      "id": 7,
      "name": "man8",
      "supercategory": ""
    },
    {
      "id": 8,
      "name": "man9",
      "supercategory": ""
    },
    {
      "id": 9,
      "name": "man10",
      "supercategory": ""
    },
    {
      "id": 10,
      "name": "man11",
      "supercategory": ""
    },
    {
      "id": 11,
      "name": "man12",
      "supercategory": ""
    },
    {
      "id": 12,
      "name": "man13",
      "supercategory": ""
    }
  ],
  "annotations": [
    {
      "area": 4167.523915361817,
      "bbox": [
        921.2998700141907,
        606.8306121826172,
        73.67040157318115,
        56.569854736328125
      ],
      "category_id": 12,
      "id": 0,
      "image_id": "171218_120213_00228",
      "iscrowd": 0,
      "segmentation": []
    },
    {
      "area": 1763.0103883534866,
      "bbox": [
        387.99999952316284,
        530.0003986358643,
        43.00032138824463,
        40.999935150146484
      ],
      "category_id": 6,
      "id": 1,
      "image_id": "171218_120213_00228",
      "iscrowd": 0,
      "segmentation": []
    },
```

**Comment by Egorundel** [comment]
And here is the folder structure:
![Screenshot from 2023-11-03 10-42-57](https://github.com/pytorch/vision/assets/65890558/d8833841-1cc5-4015-94a7-01e826d8ee74)
![Screenshot from 2023-11-03 10-42-59](https://github.com/pytorch/vision/assets/65890558/5b2fc48d-b5a1-417f-8b8d-62ec3f17ca4c)
![Screenshot from 2023-11-03 10-43-02](https://github.com/pytorch/vision/assets/65890558/3f387b25-b5b2-48c7-a5a9-f3cbf5c5a00e)
![Screenshot from 2023-11-03 10-43-19](https://github.com/pytorch/vision/assets/65890558/99aa7515-4ee5-4ad1-ba8b-93833ed94d8a)

**Comment by Egorundel** [comment]
I removed the year 2017 in the code `vision/references/detection/coco_utils.py`, since there is no year in the name in my custom dataset.

 ```python
 def get_coco(root, image_set, transforms, mode="instances", use_v2=False, with_masks=False): 
     anno_file_template = "{}_{}.json" 
     PATHS = { 
         "train": ("train", os.path.join("annotations", anno_file_template.format(mode, "train"))), 
         "val": ("val", os.path.join("annotations", anno_file_template.format(mode, "val"))), 
         # "train": ("val", os.path.join("annotations", anno_file_template.format(mode, "val"))) 
     } 
  
     img_folder, ann_file = PATHS[image_set] 
     img_folder = os.path.join(root, img_folder) 
     ann_file = os.path.join(root, ann_file)
```
   
However, my training still does not begin. What could be the problem?
![Screenshot from 2023-11-03 10-50-48](https://github.com/pytorch/vision/assets/65890558/abb6ac6a-ce32-4609-9eed-47219ddf1bf6)

**Comment by pmeier** [comment]
The `ZeroDivisionError` happens because `iterable`, i.e. the data loader, is empty. So something is wrong with the dataset. For debugging, could you try the following and post the output here

```python
from torchvision import datasets

dataset = datasets.CocoDetection(
    # fill in your values here
    "/abs/path/to/coco/images",
    "/abs/path/to/coco/annotations/instances.json",
)

print(len(dataset))
print(dataset[0])
```

**Comment by Egorundel** [comment]
@pmeier I have this situation:
![Screenshot from 2023-11-03 11-49-28](https://github.com/pytorch/vision/assets/65890558/d1d93962-a996-4960-9a1e-85b7eeb612f3)

**Comment by pmeier** [comment]
Yes, the images have to be directly in the folder that you are passing as first argument to the dataset. There is no recursive search for them.

---

Side note. Please refrain from posting screen shots. They are hard to parse. Copy the output that you have on your console and wrap them in triple backticks to get proper formatting:

````
```
<console output here>
```
````

**Comment by Egorundel** [comment]
@pmeier Hello! I took all the photos without subfolders. However, I still ran into a problem.

Test dataset in Python:
![Screenshot from 2023-11-07 16-12-27](https://github.com/pytorch/vision/assets/65890558/958e4797-6725-483f-9346-e42b50a5663a)

`torchrun` a learning:
![Screenshot from 2023-11-07 16-14-45](https://github.com/pytorch/vision/assets/65890558/ce37c02d-1305-4d07-95a8-27c4b916bf6c)

**Comment by Egorundel** [comment]
Moreover, it works almost correctly with the dataset for validation. It works for some id, but not for some:

```
(retinaNet) egorundel@egorundel-B560M-H:~$ python3
Python 3.8.10 (default, May 26 2023, 14:05:08) 
[GCC 9.4.0] on linux
>>> dataset = datasets.CocoDetection("/home/egorundel/data/NAMI_data_coco_without_subfolders/valid/images", "/home/egorundel/data/NAMI_data_coco_without_subfolders/annotations/instances_val.json")
loading annotations into memory...
Done (t=0.01s)
creating index...
index created!
>>> print(len(dataset))
2612
>>> print(dataset[0])
(<PIL.Image.Image image mode=RGB size=1280x720 at 0x7FF9BB755F10>, [{'area': 458.25967968054397, 'bbox': [771.6697853803635, 342.05004304647446, 17.169920206069946, 26.689680218696594], 'category_id': 2, 'id': 2288, 'image_id': '1', 'iscrowd': 0, 'segmentation': []}, {'area': 374.5527587470665, 'bbox': [747.8694421052933, 344.93976786732674, 13.770240545272827, 27.200160920619965], 'category_id': 2, 'id': 2289, 'image_id': '1', 'iscrowd': 0, 'segmentation': []}, {'area': 327.6760544553646, 'bbox': [679.1904270648956, 344.03004840016365, 12.84991979598999, 25.500241219997406], 'category_id': 2, 'id': 2290, 'image_id': '1', 'iscrowd': 0, 'segmentation': []}, {'area': 516.1128080655487, 'bbox': [663.2992088794708, 348.7197768688202, 18.740479946136475, 27.53999948501587], 'category_id': 2, 'id': 2291, 'image_id': '1', 'iscrowd': 0, 'segmentation': []}, {'area': 594.102719379169, 'bbox': [707.28067278862, 342.52020344138145, 23.86944055557251, 24.88967925310135], 'category_id': 2, 'id': 2292, 'image_id': '1', 'iscrowd': 0, 'segmentation': []}])
>>> print(dataset[1])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/egorundel/venvs/retinaNet/lib/python3.8/site-packages/torchvision/datasets/coco.py", line 48, in __getitem__
    image = self._load_image(id)
  File "/home/egorundel/venvs/retinaNet/lib/python3.8/site-packages/torchvision/datasets/coco.py", line 40, in _load_image
    path = self.coco.loadImgs(id)[0]["file_name"]
  File "/home/egorundel/venvs/retinaNet/lib/python3.8/site-packages/pycocotools/coco.py", line 227, in loadImgs
    return [self.imgs[id] for id in ids]
  File "/home/egorundel/venvs/retinaNet/lib/python3.8/site-packages/pycocotools/coco.py", line 227, in <listcomp>
    return [self.imgs[id] for id in ids]
KeyError: '0'
>>> 

```

**Comment by NicolasHug** [comment]
Hi @Egorundel ,

I'm really sorry but as mentioned already in previous threads, this is a bug tracker, not a place for user-help. Please consider posting usage quetsions on https://discuss.pytorch.org/ instead.
